internal/subscriber/retry/delayconsumer_test.go

```diff
@@ -81,7 +95,162 @@ func TestDelayConsumer_Run(t *testing.T) {
+				func(arg0 context.Context, arg1 messagebroker.GetMessagesFromTopicRequest) (*messagebroker.GetMessagesFromTopicResponse, error) {
+					if count == 0 {
+						count++
+						return &messagebroker.GetMessagesFromTopicResponse{Messages: tt.args.messages}, nil
+					} else {
+						return &messagebroker.GetMessagesFromTopicResponse{}, nil
+					}
+				}).AnyTimes()
+			go dc.Run(ctx)
+			select {
+			case err := <-errCh:
+				if !tt.wantErr {
+					t.Errorf("Got Error : %v", err)
+				}
+			case <-time.NewTicker(time.Millisecond * 500).C:
+				assert.Equal(t, len(dc.cachedMsgs), 1)
+				assert.True(t, reflect.DeepEqual(dc.cachedMsgs[0], tt.expected))
+			}
+			cancel()
+			<-time.NewTicker(time.Millisecond * 1).C
+		})
+	}
+}
+
+func TestDelayConsumer_Run_DeadLetter(t *testing.T) {
+	ctx, ctrl, subCtx, cancel, brokerstore, handler, consumer, cache := setupDC(t)
+	subs := getDummySubscriptionModel()
+	subscriberID := "subscriber-dl-id"
+	errCh := make(chan error)
+	dlqTopicChan := make(chan *messagebroker.SendMessageToTopicRequest)
+	producer := buildMockDLQProducer(ctx, ctrl, dlqTopicChan)
+	brokerstore.EXPECT().GetProducer(gomock.AssignableToTypeOf(subCtx), messagebroker.ProducerClientOptions{
+		Topic:     "dlt1",
+		TimeoutMs: 100,
+	}).Return(producer, nil).AnyTimes()
+	dc, err := NewDelayConsumer(subCtx, subscriberID, "t1-dl", subs, brokerstore, handler, cache, errCh)
+	assert.NotNil(t, dc)
+	assert.Nil(t, err)
+	assert.NotNil(t, dc.LogFields())
+	tFut := time.Now().Add(time.Second * 10)
+	msgs := make([]messagebroker.ReceivedMessage, 0)
+	msg := getDummyBrokerMessageDLQ()
+	msg.NextDeliveryTime = tFut
+	msgs = append(msgs, msg)
+	count := 0
+	type args struct {
+		messages []messagebroker.ReceivedMessage
+	}
+	tests := []struct {
+		name     string
+		wantErr  bool
+		expected *messagebroker.SendMessageToTopicRequest
+		args     args
+	}{
+		{
+			name: "Test Delay Consumer msg pushed to DLQ success",
+			args: args{
+				messages: msgs,
+			},
+			expected: getMockSendMessageToTopicRequest(),
+			wantErr:  false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			consumer.EXPECT().ReceiveMessages(gomock.AssignableToTypeOf(subCtx), messagebroker.GetMessagesFromTopicRequest{
+				NumOfMessages: 10,
+				TimeoutMs:     100,
+			}).DoAndReturn(
+				func(arg0 context.Context, arg1 messagebroker.GetMessagesFromTopicRequest) (*messagebroker.GetMessagesFromTopicResponse, error) {
+					if count == 0 {
+						count++
+						return &messagebroker.GetMessagesFromTopicResponse{Messages: tt.args.messages}, nil
+					} else {
+						return &messagebroker.GetMessagesFromTopicResponse{}, nil
+					}
+				}).AnyTimes()
+			go dc.Run(ctx)
+			for {
+				select {
+				case err := <-errCh:
+					t.Errorf("Got Error : %v", err)
+				case req := <-dlqTopicChan:
+					assert.Equal(t, req, tt.expected)
+				case <-time.NewTicker(time.Millisecond * 500).C:
+					cancel()
+					<-time.NewTicker(time.Millisecond * 1).C
+					return
```

Comment:
why do we need a for select case here, and not just select case..?
Also, don't add ticker to wait on cancel, check for a signal on dc done channel